## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [510]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [511]:
# View the column names in the data
column_names = df.columns
print(column_names)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [512]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [513]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.value_counts
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [514]:

# Calculate average unit_price, unit_cost, and qty per client_id
average_per_client = df.groupby('client_id').agg({
    'unit_price': 'mean',
    'unit_cost': 'mean',
    'qty': 'mean'
})

# Get the top 5 clients based on the mean qty (or any relevant metric)
top_5_clients = average_per_client['qty'].nlargest(5).index

# Get the average values for the top 5 clients
average_top_5_clients = average_per_client.loc[top_5_clients]

# Calculate overall average for comparison
overall_average = df[['unit_price', 'unit_cost', 'qty']].mean()

# Print the average values for all clients and the top 5 clients, and compare with the overall averages
print("Average unit_price, unit_cost, and qty per client_id:")
print(average_per_client)

print("\nAverage values for the top 5 clients:")
print(average_top_5_clients)

print("\nOverall average values for comparison:")
print(overall_average)


Average unit_price, unit_cost, and qty per client_id:
           unit_price   unit_cost          qty
client_id                                     
10033      192.186842  135.644737   165.947368
10091       79.596429   51.322143  2122.214286
10218      104.889643   80.323214  1162.821429
10351      121.038800   87.309000   152.300000
10435      110.529301   82.391538   258.510490
...               ...         ...          ...
99400      138.387101   98.532609   331.768116
99450      100.481739   70.372174   386.521739
99736      126.580000   88.659000  1269.740000
99973      126.900000   92.408000   144.925000
99984      139.719394  105.244545  1455.575758

[999 rows x 3 columns]

Average values for the top 5 clients:
           unit_price   unit_cost           qty
client_id                                      
30143      156.838182  121.629091  25638.454545
78965      131.079306   94.916821  23313.884393
41656      157.324375  110.238125  15149.093750
85931      138.720545  101.68000

In [515]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print("Top three item categories with the most entries:")
print(top_categories)


Top three item categories with the most entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [516]:
# For the category with the most entries and show number of entries per subcategory
# which subcategory had the most entries?
most_entries_subcategory = subcategory_counts.idxmax()
top_category_df = df[df['category'] == top_category]
subcategory_counts = top_category_df['subcategory'].value_counts()
ost_entries_subcategory = subcategory_counts.idxmax()
count_most_entries_subcategory = subcategory_counts.max()
print(f"The subcategory with the most entries in the top category '{top_category}' is '{most_entries_subcategory}'")
print(f"Number of entries for the '{most_entries_subcategory}' subcategory: {count_most_entries_subcategory}")


The subcategory with the most entries in the top category 'consumables' is 'bathroom supplies'
Number of entries for the 'bathroom supplies' subcategory: 6424


In [531]:
# Which five clients had the most entries in the data?

top_clients = df['client_id'].value_counts().head(5)

print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [518]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5)
top_5_client_ids = top_clients.index.tolist()

print("Client IDs of the top 5 clients:")
print(top_5_client_ids)


Client IDs of the top 5 clients:
[33615, 66037, 46820, 38378, 24741]


In [519]:
# How many total units (the qty column) did the
# client with the most entries order order?
most_common_client = df['client_id'].value_counts().idxmax()
filtered_df = df[df['client_id'] == most_common_client]
total_units_ordered = filtered_df['qty'].sum()
print(f"The client with the most entries ordered a total of {total_units_ordered} units.")

The client with the most entries ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [520]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
selected_columns = ['unit_price', 'qty', 'subtotal']
subset_df = df[selected_columns]

print(subset_df.head())

   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00


In [521]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
selected_columns = ['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']
subset_df = df[selected_columns]

print(subset_df.head())



   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20


In [522]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df['total_price'] = df['subtotal'] + df['shipping_price']
df['total_price_with_tax'] = df['total_price'] * (1 + sales_tax_rate)
df['total_price_with_tax'] = df['total_price_with_tax'].round(2)
selected_columns = ['subtotal', 'shipping_price', 'total_price_with_tax']
subset_df = df[selected_columns]

print(subset_df.head())


    subtotal  shipping_price  total_price_with_tax
0  115164.00         5512.50             131839.08
1     523.95          312.90                914.26
2     527.28          458.64               1077.12
3    1056.18          356.70               1543.57
4    3902.00         6500.20              11364.40


In [523]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
if 'total_price' in df.columns:
    df = df.drop(columns=['total_price'])

print(df.head(3))


     first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number   subtotal  total_weight  shipping_price  total_price_with_tax  \
0            1  115164.00        787.50         5512.50             131839.08   
1            0 

In [524]:
# Create a column for the profit of
# each line using line cost and line price
df['line_profit'] = df['total_price_with_tax'] - df['line_cost']
print(df.head(3))


     first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

    subtotal  total_weight  shipping_price  total_price_with_tax  line_cost  \
0  115164.00        787.50         5512.50             131839.08   85597.05   
1     523.9

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [525]:
# Check your work using the totals above
order_totals = df.groupby('order_id')['total_price_with_tax'].sum().reset_index()
provided_order_ids = [2742071, 2173913, 6128929]
for order_id in provided_order_ids:
    total = order_totals.loc[order_totals['order_id'] == order_id, 'total_price_with_tax'].values
    print(f"order {order_id} total: ${total[0] if len(total) > 0 else 0:.2f}")


order 2742071 total: $152811.89
order 2173913 total: $162388.71
order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [526]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
client_totals = df.groupby('client_id')['total_price_with_tax'].sum()
top_5_clients = df['client_id'].value_counts().head(5).index
for client_id in top_5_clients:
    total_spending = client_totals.get(client_id, 0)
    print(f"clinet_id {client_id} spent: ${total_spending:.2f}")


clinet_id 33615 spent: $8377308.52
clinet_id 66037 spent: $10259514.79
clinet_id 46820 spent: $9743794.36
clinet_id 38378 spent: $12906550.87
clinet_id 24741 spent: $82268892.02


In [527]:
# Calculate total spending for the top 5 clients by quantity
top_5_clients = df['client_id'].value_counts().head(5).index
total_spending_top_5 = df[df['client_id'].isin(top_5_clients)]['total_price_with_tax'].sum() / 1_000_000  # Convert to millions

# Calculate total spending for the top 20% of clients by spending
total_spending_all = df.groupby('client_id')['total_price_with_tax'].sum()
top_20_percent_clients = total_spending_all.nlargest(int(len(total_spending_all) * 0.2)).index
total_spending_top_20_percent = df[df['client_id'].isin(top_20_percent_clients)]['total_price_with_tax'].sum() / 1_000_000  # Convert to millions

# Calculate total spending for all clients
total_spending_all = total_spending_all.sum() / 1_000_000  # Convert to millions

# Calculate average spending for all clients
average_spending_all = total_spending_all / df['client_id'].nunique()

# Print the total spending of top 5 clients, total spending for top 20% of clients, total spending for all clients,
# and average spending of all clients in millions
print(f"Total spending of top 5 clients by quantity: ${total_spending_top_5:.2f}M")
print(f"Total spending of top 20% of clients by spending: ${total_spending_top_20_percent:.2f}M")
print(f"Total spending for all clients: ${total_spending_all:.2f}M")
print(f"Average spending of all clients: ${average_spending_all:.2f}M")



Total spending of top 5 clients by quantity: $123.56M
Total spending of top 20% of clients by spending: $4137.96M
Total spending for all clients: $5642.67M
Average spending of all clients: $5.65M


In [528]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

specific_clients = df[df['client_id'].isin(top_5_clients)]
client_summary = specific_clients.groupby('client_id').agg({
    'qty': 'sum',                     
    'shipping_price': 'sum',           
    'total_price_with_tax': 'sum',
    'line_cost': 'sum',      
    'line_profit': 'sum'  
}).sort_values('line_profit', ascending=False)

print(client_summary)


              qty  shipping_price  total_price_with_tax    line_cost  \
client_id                                                              
24741      239862      5126448.37           82268892.02  45688899.71   
38378       73667      3429455.40           12906550.87   9634720.98   
66037       43018      1395151.85           10259514.79   7004482.98   
46820       75768      1601448.84            9743794.36   7007191.64   
33615       64313      1828984.89            8377308.52   6175313.91   

           line_profit  
client_id               
24741      36579992.31  
38378       3271829.89  
66037       3255031.81  
46820       2736602.72  
33615       2201994.61  


In [529]:
# Grouping and aggregating data
specific_clients = df[df['client_id'].isin(top_5_clients)]
client_summary = specific_clients.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price_with_tax': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
}) 

# Convert monetary columns to millions with "$" symbol and format to 2 decimal places
monetary_cols = ['shipping_price', 'total_price_with_tax', 'line_profit', 'line_cost']
client_summary[monetary_cols] /= 1_000_000
client_summary[monetary_cols] = client_summary[monetary_cols].applymap('${:.2f}M'.format)

# Rename columns for presentation
client_summary.columns = [
    'Total Units (M)',
    'Shipping (M)',
    'Total Revenue (M)',
    'Total Cost (M)',
    'Total Profit (M)'
]

# Reorder columns
client_summary = client_summary[['Total Units (M)', 'Shipping (M)', 'Total Revenue (M)', 'Total Cost (M)', 'Total Profit (M)']]

print("Formatted Summary DataFrame for the top 5 clients in millions of dollars:")
print(client_summary)


Formatted Summary DataFrame for the top 5 clients in millions of dollars:
           Total Units (M) Shipping (M) Total Revenue (M) Total Cost (M)  \
client_id                                                                  
24741               239862       $5.13M           $82.27M        $45.69M   
33615                64313       $1.83M            $8.38M         $6.18M   
38378                73667       $3.43M           $12.91M         $9.63M   
46820                75768       $1.60M            $9.74M         $7.01M   
66037                43018       $1.40M           $10.26M         $7.00M   

          Total Profit (M)  
client_id                   
24741              $36.58M  
33615               $2.20M  
38378               $3.27M  
46820               $2.74M  
66037               $3.26M  


In [530]:
# Sort the updated data by "Total Profit" form highest to lowest
client_summary = client_summary.sort_values('Total Profit (M)', ascending=False)

print("Formatted Summary DataFrame for the top 5 clients sorted by Total Profit:")
print(client_summary)

Formatted Summary DataFrame for the top 5 clients sorted by Total Profit:
           Total Units (M) Shipping (M) Total Revenue (M) Total Cost (M)  \
client_id                                                                  
24741               239862       $5.13M           $82.27M        $45.69M   
38378                73667       $3.43M           $12.91M         $9.63M   
66037                43018       $1.40M           $10.26M         $7.00M   
46820                75768       $1.60M            $9.74M         $7.01M   
33615                64313       $1.83M            $8.38M         $6.18M   

          Total Profit (M)  
client_id                   
24741              $36.58M  
38378               $3.27M  
66037               $3.26M  
46820               $2.74M  
33615               $2.20M  


# Summary
The combined spending of the top 5 clients amounts to $123.56 million, constituting approximately 2% of the total expenditure of $5,642.67 million across all clients. Notably, their spending significantly surpasses the average client expenditure of $5.6 million. Delving deeper, the top 20% of clients collectively contributed $4,137.96 million, representing a substantial 73% of the total expenditure.

It is evident that directing attention towards not only the top 5 but also the top 20% of clients is imperative for the company, considering they generate the majority (73%) of the business. Focusing efforts on these client segments would prove beneficial for the company's growth and revenue.

